docs/lang/control-and-nondeterminism.md

```diff
@@ -0,0 +1,509 @@
+made wrong choices on the way. If we like to enumerate all possible state
+successors, like TLC does, we have to backtrack. In general, the course of
+action depends on the program analysis that you implement. For instance,
+a random simulator could simply backtrack and randomly choose another value.
+
+<a name="nondetExists"></a>
+### Non-determinism in `\E x \in S: P`
+
+We only have to consider the following case: `\E x \in S: P` is evaluated in a
+partial state `s`, and there is a primed state variable `y'` that satisfies two
+conditions:
+
+ 1. The predicate `P` refers to `y'`, that is, `P` has to assign a value to `y'`.
+ 2. The value of `y'` is not defined yet, that is, `s.y' = Null`.
+
+If the above assumptions do not hold true, the expression `\E x \in S: P` does
+not have non-determinism and it can be evaluated by following the standard
+deterministic semantics of exists, see [[Logic]](./logic.md).
+
+Now it is very easy to evaluate `\E x \in S: P`. We simply evaluate the
+following expression:
+
+```tla
+  LET x == GUESS S IN P
+```
+
+It is the job of `GUESS S` to tell us what value of `x` should be
+evaluated. There are three possible outcomes:
+
+ 1. Predicate `P` evaluates to `TRUE` when using the provided value of `x`.
+ In this case, `P` assigns the value of an expression `e` to `y'` as soon as
+ the evaluator meets the expression `y' = e`.
+ The evaluation may continue.
+ 1. Predicate `P` evaluates to `FALSE` when using the provided value of `x`.
+ Well, that was a wrong guess. According to our semantics, the evaluation
+ halts. See the above discussion on "halting".
+ 1. The set `S` is empty, and `GUESS S` halts.  See the above discussion on
+ "halting". 
+
+**Example.** Consider the following specification:
+
+```tla
+VARIABLE x
+Init == x = 0
+Next ==
+  \E i \in Int:
+    i > x /\ x' = i
+```
+
+It is easy to evaluate `Init`: It does not contain non-determinism and it
+evaluates to the state `[x |-> 0]`. When evaluating `Next` in the partial state
+`[x |-> 0, x' |-> Null]`, we have plenty of choices. Actually, we have
+infinitely many choices, as the set `Int` is infinite.  TLC would immediately
+fail here. But there is no reason for our evaluation to fail.  Simply ask the
+oracle. Below we give three examples of how the evaluation works:
+
+```
+1. (GUESS Int) returns 10. (LET i == 10 IN i > x /\ x' = i) is TRUE, x' is assigned 10.
+2. (GUESS Int) returns 0. (LET i == 0 IN i > x /\ x' = i) is FALSE. Halt.
+3. (GUESS Int) returns -20. (LET i == -20 IN i > x /\ x' = i) is FALSE. Halt.
+```
+
+<a name="nondetOr"></a>
+### Non-determinism in disjunctions
+
+Consider a disjunction that comprises `n` clauses:
+
+```tla
+  \/ P_1
+  \/ P_2
+  ...
+  \/ P_n
+```
+
+Assume that we evaluate the disjunction in a partial state `s`. Further,
+let us say that `Unassigned(s)` is the set of state variables that evaluate
+to `Null` in `s`. For every `P_i` we construct the set of state variables 
```

Comment:
Again, find a better way to differentiate between `s.x` and `s_Next.x` (aka `s.x'`). This absorption of primed variable names into partial states is very confusing at times.